lawlibrary/views/legislation.py

```diff
@@ -0,0 +1,121 @@
+                "localities": [
+                    {
+                        "code": locality.code,
+                        "name": locality.name,
+                    }
+                    for locality in Locality.objects.filter(
+                        jurisdiction__name=jurisdiction
+                    )
+                ],
+            }
+            provincial_legislation_list.append(locality_dict)
+
+        context["provincial_legislation_list"] = provincial_legislation_list
+        context["facet_data"] = {
+            "years": self.get_years(),
+            "taxonomies": self.get_taxonomies(),
+        }
+
+        return context
+
+    def get_jurisdictions(self):
+        return (
+            self.get_queryset()
+            .order_by("jurisdiction__name")
+            .values_list("jurisdiction__name", flat=True)
+            .distinct()
+        )
+
+    def get_queryset(self):
+        return super().get_queryset()
+
+    def get_years(self):
+        qs = (
+            self.get_queryset()
+            .filter(locality=None)
+            .order_by()
+            .values_list("date__year", flat=True)
+            .distinct()
+        )
+        return sorted(qs, reverse=True)
+
+    def get_taxonomies(self):
+        taxonomies = []
+        for legislation in self.get_queryset().filter(locality=None):
+            for taxonomy in legislation.taxonomies.all():
+                if taxonomy.topic.name not in taxonomies:
+                    taxonomies.append(taxonomy.topic.name)
+        return taxonomies
+
+
+class ProvincialLegislationListView(ListView):
+    model = Legislation
+    template_name = "lawlibrary/provincial_legislation_list.html"
+    context_object_name = "documents"
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+
+        qs = (
+            self.get_queryset()
+            .filter(locality__code=self.kwargs["code"])
+            .distinct("work_frbr_uri")
+            .order_by("work_frbr_uri", "-date")
+        )
+
+        context["legislation_table"] = LegislationSerializer(qs, many=True).data
+        context["facet_data"] = {
+            "years": self.get_years(),
+            "taxonomies": self.get_taxonomies(),
+        }
+
+        return context
+
+    def get_queryset(self):
+        return super().get_queryset()
+
+    def get_years(self):
+        qs = (
+            self.get_queryset()
+            .filter(locality__code=self.kwargs["code"])
+            .order_by()
+            .values_list("date__year", flat=True)
+            .distinct()
+        )
+        return sorted(qs, reverse=True)
+
+    def get_taxonomies(self):
+        taxonomies = []
+        for legislation in self.get_queryset().filter(
+            locality__code=self.kwargs["code"]
+        ):
+            for taxonomy in legislation.taxonomies.all():
+                if taxonomy.topic.name not in taxonomies:
+                    taxonomies.append(taxonomy.topic.name)
```

Comment:
You can replace these for loops with `self.get_queryset().filter(locality__code=self.kwargs["code"]).values_list('taxonomies__topic__name', flat=True)`